### ISSUE TITLE
Commit breaks CUDA9 builds [NativeFunctions: support backend-specific dispatch]

### ISSUE THREAD

**Issue by VladislavZavadskyy** [pytorch/pytorch#3807]
This one: 445cc1f5b9a1647949c7b27b60e9800339f88db9 
Here's the [log](https://gist.github.com/VladislavZavadskyy/f587dbbe68469d005c807ab98946533f). And the setup is: Ubuntu 17.10, gcc 6.4.0, cuda 9.0 and cudnn 7.0.3.
Hard resetting to previous commit makes build go smooth, so I'm pretty sure it's the one.

**Comment by sdmonov** [comment]
Can you try running setup.py again after you see this error. I see same error and when retry the setup.py for second time it works for me. I guess this is just temporary workaround until the problem is fixed.

**Comment by vadimkantorov** [comment]
Current version has a subtle bug that makes build from sources pick up headers of a previous installed version: https://github.com/pytorch/pytorch/issues/3669

You may try removing completely the existing install, and start over (if haven't tried yet).

**Comment by peterjc123** [comment]
A temporary fix for that would be calling the gen.py before the calling setup.py. That is:
```Bash
mkdir torch\lib\build\ATen\src\ATen
cd torch\lib\build\ATen\src\ATen
mkdir ATen
python  ../../../../../../aten/src/ATen/gen.py -s  ../../../../../../aten/src/ATen  ../../../../../../aten/src/ATen/Declarations.cwrap  ../../../../../../aten/src/THNN/generic/THNN.h  ../../../../../../aten/src/THCUNN/generic/THCUNN.h  ../../../../../../aten/src/ATen/nn.yaml  ../../../../../../aten/src/ATen/native/native_functions.yaml
cd ../../../../../..
```
Also a PR for that is #3757, which is not so clean but works.

**Comment by napsternxg** [comment]
I am facing similar issue. Reported here: #3822 

Following @peterjc123 suggested steps, I am still getting the same error. 
```
mkdir -p torch/lib/build/ATen/src/ATen/ATen
cd torch/lib/build/ATen/src/ATen
python  ../../../../../../aten/src/ATen/gen.py -s  ../../../../../../aten/src/ATen  ../../../../../../aten/src/ATen/Declarations.cwrap  ../../../../../../aten/src/THNN/generic/THNN.h  ../../../../../../aten/src/THCUNN/generic/THCUNN.h  ../../../../../../aten/src/ATen/nn.yaml  ../../../../../../aten/src/ATen/native/native_functions.yaml
cd ../../../../../..
python setup.py install
```

@vadimkantorov @soumith any suggested workarounds ?

**Comment by vadimkantorov** [comment]
@peterjc123 The PR https://github.com/pytorch/pytorch/pull/3757 only adds a hack for the Windows platform, right?

If you're also referring to the problem of using obsolete includes, for a proper fix, I think, one should try to figure out how to fix headers before system headers. At least that is the problem in https://github.com/pytorch/pytorch/issues/3669

**Comment by HenryJia** [comment]
Any updates on this bug? I still can't seem to find a workaround?

**Comment by mikael10j** [comment]
Hi,
it seems to be related to gcc 6. I had the same issue with Ubuntu 17.04, gcc 6.3 and CUDA 9.0.
I could to compile with gcc 5.4.1, but without gloo (NO_Distributed=1), otherwise it fails.

**Comment by HenryJia** [comment]
Compiling with gcc 5.4.0-1 leads to "error: unrecognized command line option ‘-fno-plt’" for me
I'm on Arch Linux

**Comment by soumith** [comment]
@HenryJia `fno-plt` seems to be an Arch + CUDA thing. See https://www.reddit.com/r/archlinux/comments/6nxxre/build_error_gcc5_fail_on_fnoplt_flag/

**Comment by HenryJia** [comment]
@soumith Well, simply doing gcc-5 -fno-plt is enough to raise that error, I'm not sure how I'm supposed to get around this?

**Comment by soumith** [comment]
@HenryJia from reading that reddit / archlinux thread it is like this:
- gcc-5 doesn't have `fno-plt`.
- archlinux sets it's CFLAGS to have `fno-plt` because it assumes default compiler is gcc 6+

**Comment by HenryJia** [comment]
@soumith Hmmm, OK, but how do I stop it from doing that? my $CFLAGS and $CXXFLAGS are empty?

**Comment by soumith** [comment]
i am not entirely sure, I dont have archlinux. Try asking on some of the Arch forums, they must've figured it out by now.

**Comment by HenryJia** [comment]
Alright, will do

**Comment by ppwwyyxx** [comment]
ArchLinux w/ gcc5, able to build successfully by the following change:
```diff
diff --git i/setup.py w/setup.py
index e484692f..75e1836e 100644
--- i/setup.py
+++ w/setup.py
@@ -90,7 +90,7 @@ import distutils.sysconfig
 cfg_vars = distutils.sysconfig.get_config_vars()
 for key, value in cfg_vars.items():
     if type(value) == str:
-        cfg_vars[key] = value.replace("-Wstrict-prototypes", "")
+        cfg_vars[key] = value.replace("-Wstrict-prototypes", "").replace("-fno-plt", "")

 ################################################################################
 # Custom build commands
```

**Comment by HenryJia** [comment]
adding CFLAGS="${CFLAGS/-fno-plt/}"  CXXFLAGS="${CXXFLAGS/-fno-plt/}" to the build command also seems to do the trick. I think it essentially does the same thing

**Comment by nmilosev** [comment]
Got similar error on Fedora 28.

Got it to build though:

- pytorch 0.4.0
- gcc 4.9.3 (https://copr.fedorainfracloud.org/coprs/davidva/gcc49/)
- cuda 9.1 and nvidia drivers from negativo repository (optimus laptop) - note here for negativo's cuda https://github.com/torch/torch7/issues/1138#issuecomment-381380180
- build without `NO_Distributed=1`

nvidia-smi doesn't report a process running but I see the speedup compared to CPU and also memory usage in nvidia-smi. This is probably optimus related.

Another issue is that if I try to use prebuilt pytorch binaries I run into this: https://discuss.pytorch.org/t/runtimeerror-cudnn-status-mapping-error/4370

Thanks for the useful information in this thread! <3

**Comment by SsnL** [comment]
This is an NVIDIA NVCC bug, which is fixed in CUDA 9.2.

closed via #8863 which adds a clear error message
